backend/de.metas.business/src/main/java/de/metas/costrevaluation/process/M_CostRevaluation_CreateLines_Process.java

```diff
@@ -0,0 +1,136 @@
+import de.metas.costrevaluation.impl.CostRevaluationId;
+import de.metas.costrevaluation.impl.CostRevaluationLine;
+import de.metas.process.IProcessPrecondition;
+import de.metas.process.IProcessPreconditionsContext;
+import de.metas.process.JavaProcess;
+import de.metas.process.ProcessPreconditionsResolution;
+import de.metas.product.ProductId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.ad.dao.IQueryBL;
+import org.compiere.SpringContextHolder;
+import org.compiere.model.I_M_CostRevaluation;
+
+import java.util.List;
+
+import static org.adempiere.model.InterfaceWrapperHelper.loadOutOfTrx;
+
+/**
+ * Process to create M_CostRevaluationLine records for products stocked based on accounting schema and cost element of M_CostRevaluation
+ */
+public class M_CostRevaluation_CreateLines_Process extends JavaProcess implements IProcessPrecondition
+{
+	private final IQueryBL queryBL = Services.get(IQueryBL.class);
+	private final CostRevaluationRepository costRevaluationRepo = SpringContextHolder.instance.getBean(CostRevaluationRepository.class);
+	private final ICurrentCostsRepository currentCostsRepo = SpringContextHolder.instance.getBean(ICurrentCostsRepository.class);
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(final @NonNull IProcessPreconditionsContext context)
+	{
+		if (context.isNoSelection())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNoSelection().toInternal();
+		}
+		else if (context.isMoreThanOneSelected())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNotSingleSelection().toInternal();
+		}
+		else if (!isDraftDocument(context))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Already in progress/finished.");
+		}
+		return ProcessPreconditionsResolution.accept();
+	}
+
+	private boolean isDraftDocument(final @NonNull IProcessPreconditionsContext context)
+	{
+		final CostRevaluationId costRevaluationId = CostRevaluationId.ofRepoId(context.getSingleSelectedRecordId());
+		return costRevaluationRepo.isDraftedDocument(costRevaluationId);
+	}
+
+	@Override
+	protected String doIt() throws Exception
+	{
+
+		final ImmutableSet<ProductId>  products = fetchProducts();
+		if (products.isEmpty()) return "NoSelection@";
+
+		final CostRevaluationId costRevaluationId = CostRevaluationId.ofRepoId(getRecord_ID());
+		final I_M_CostRevaluation costRevaluation = loadOutOfTrx(costRevaluationId, I_M_CostRevaluation.class);
+		final AcctSchemaId acctSchemaId = AcctSchemaId.ofRepoId(costRevaluation.getC_AcctSchema_ID());
+		final CostElementId costElementId = CostElementId.ofRepoId(costRevaluation.getM_CostElement_ID());
+
+
+
+		for (ProductId productId : products)
+		{
+			createCostRevaluationLineForProduct(acctSchemaId, costElementId, productId);
+		}
+
+		return "@OK@";
+	}
+
+	private void createCostRevaluationLineForProduct(@NonNull final AcctSchemaId acctSchemaId,
```

Comment:
avoid SQL N+1 issue.

consider taking productIds set as parameter.

and then, 
* when you fetch current cost prices, fetch them all together by productIds
* when you save the lines save them all together, so the method which checks if the line record exists, will check only once for all lines